clang/lib/CAS/IncludeTree.cpp

```diff
@@ -0,0 +1,210 @@
+
+using namespace clang;
+using namespace clang::cas;
+
+template <typename NodeT>
+Expected<NodeT> IncludeTreeBase<NodeT>::create(CASDB &DB,
+                                               ArrayRef<ObjectRef> Refs,
+                                               ArrayRef<char> Data) {
+  SmallString<256> Buf{NodeT::getNodeKind()};
+  Buf.reserve(Data.size() + NodeT::getNodeKind().size());
+  Buf.append(Data.begin(), Data.end());
+  auto Node = DB.store(Refs, Buf);
+  if (!Node)
+    return Node.takeError();
+  return NodeT(ObjectProxy::load(DB, *Node));
+}
+
+Expected<IncludeFile> IncludeFile::create(CASDB &DB, StringRef Filename,
+                                          ObjectRef Contents) {
+  auto PathHandle = DB.storeFromString({}, Filename);
+  if (!PathHandle)
+    return PathHandle.takeError();
+  std::array<ObjectRef, 2> Refs{DB.getReference(*PathHandle), Contents};
+  return IncludeTreeBase::create(DB, Refs, {});
+}
+
+llvm::Error IncludeTree::forEachInclude(
+    llvm::function_ref<llvm::Error(std::pair<IncludeTree, unsigned>)>
+        Callback) {
+  size_t RefI = 0;
+  return forEachReference([&](ObjectRef Ref) -> llvm::Error {
+    if (RefI == 0) {
+      ++RefI;
+      return llvm::Error::success();
+    }
+    size_t IncludeI = RefI - 1;
+    ++RefI;
+    auto Include = getInclude(Ref);
+    if (!Include)
+      return Include.takeError();
+    return Callback({*Include, getIncludeOffset(IncludeI)});
+  });
+}
+
+template <typename T>
+static void addToBuffer(T Value, SmallVectorImpl<char> &Buffer,
+                        llvm::support::endianness Endian) {
+  Buffer.resize(Buffer.size() + sizeof(T));
+  llvm::support::endian::write<T>(Buffer.end() - sizeof(T), Value, Endian);
+}
+
+Expected<IncludeTree>
+IncludeTree::create(CASDB &DB, SrcMgr::CharacteristicKind FileCharacteristic,
+                    ObjectRef IncludedBy,
+                    ArrayRef<std::pair<ObjectRef, unsigned>> Includes,
+                    llvm::SmallBitVector Checks) {
+  // The data buffer is composed of
+  // 1. `unsigned` offsets of includes
+  // 2. 1 byte for `CharacteristicKind`
+  // 3. variable number of bitset bytes for `Checks`.
+
+  char Kind = FileCharacteristic;
+  assert(Kind == FileCharacteristic && "SrcMgr::CharacteristicKind too big!");
+  assert(IncludeFile::isValid(DB, IncludedBy));
+  SmallVector<ObjectRef, 16> Refs;
+  Refs.reserve(Includes.size() + 1);
+  Refs.push_back(IncludedBy);
+  SmallString<64> Buffer;
+  Buffer.reserve(Includes.size() * sizeof(unsigned) + 1);
+
+  for (const auto &Include : Includes) {
+    ObjectRef FileRef = Include.first;
+    unsigned Offset = Include.second;
+    assert(IncludeTree::isValid(DB, FileRef));
+    Refs.push_back(FileRef);
+    addToBuffer(Offset, Buffer, llvm::support::native);
+  }
+
+  Buffer += Kind;
+
+  uintptr_t Store;
+  ArrayRef<uintptr_t> BitWords = Checks.getData(Store);
+  size_t RemainingBitsCount = Checks.size();
+  while (RemainingBitsCount > 0) {
```

Comment:
Is there any reason not to just write all of `getData()`? It would be simpler than this, and it looks like unused bits should be 0 (you already kind of depend on this below to avoid unstable output since you write the final byte without masking unused bits).

Reply:
I didn't want to waste 8 bytes if only a few checks existed, I suspect the common case is that this takes only 1 byte.

Reply:
Oh right, this is only for has_include. Makes sense.